specutils/spectra/spectrum1d.py

```diff
@@ -18,53 +15,52 @@ class Spectrum1D(OneDSpectrumMixin, NDDataRef):
                  spectral_axis_unit=None, *args, **kwargs):
-
-        if not isinstance(flux, Quantity):
-            flux = Quantity(flux, unit=unit or "Jy")
-
-        if spectral_axis is not None and not isinstance(spectral_axis, Quantity):
-            spectral_axis = Quantity(spectral_axis, unit=spectral_axis_unit or u.AA)
-
-        # If spectral_axis had not been defined, attempt to use the wcs
-        # information, if it exists
-        if spectral_axis is None and wcs is not None:
-            if isinstance(wcs, WCS):
-
-                # Try to reference the spectral axis
-                wcs_spec = wcs.sub([WCSSUB_SPECTRAL])
-
-                # Check to see if it actually is a real coordinate description
-                if wcs_spec.naxis == 0:
-                    # It's not real, so attempt to get the spectral axis by
-                    # specifying axis by integer
-                    wcs_spec = wcs.sub([wcs.naxis])
-
-                # Construct the spectral_axis array
-                spectral_axis = wcs_spec.all_pix2world(
-                    np.arange(flux.shape[0]), 0)[0]
-
-                # Try to get the spectral_axis unit information
-                try:
-                    spectral_axis_unit = wcs.wcs.cunit[0]
-                    if spectral_axis_unit == dimensionless_unscaled:
-                        logging.warning("Spectral_axis unit is dimensionless. Pls set one.")
-                except AttributeError:
-                    logging.warning("No spectral_axis unit information in WCS.")
-                    spectral_axis_unit = Unit("")
-
-                spectral_axis = spectral_axis * spectral_axis_unit
-
-                if wcs.wcs.restfrq != 0:
-                    self._rest_value = wcs.wcs.restfrq * u.Hz
-                elif wcs.wcs.restwav != 0:
-                    self._rest_value = wcs.wcs.restwav * u.AA
+        # Attempt to parse the WCS. If not WCS object is given, try instead to
+        # parse a given wavelength array. This is put into a GWCS object to
+        # then be used behind-the-scenes for all specutils operations.
+        if wcs is not None:
+            wcs = WCSWrapper(wcs)
+        elif spectral_axis is not None:
+            spectral_axis = u.Quantity(spectral_axis,
+                                       unit=spectral_axis_unit)
+
+            wcs = WCSWrapper.from_array(spectral_axis)
+        else:
+            # If not wcs and not spectral axis has been given, raise an error
+            raise LookupError("No WCS object or spectral axis information has "
+                              "been given. Please provide one.")
+
+        if not isinstance(flux, u.Quantity):
+            flux = u.Quantity(flux, unit=unit or "Jy")
+
+        self._velocity_convention = None
+
+        # Currently, only a fits wcs object stores the rest wavelength or
+        # frequency information in the wcs object. In any other case, the user
+        # will be given an error to provide these.
```

Comment:
Oops, technically the comment is out of date. The GWCS object should simply log a warning telling the user to set the `rest_value` explicitly on the `Spectrum1D` object, since the GWCS wrapper provides properties (that simply return 0).